### PR TITLE
XDV-05 also reduce ownerCount if the last candidate is deleted

### PR DESCRIPTION
**Pull request by wanwiset25**
# Proposed changes
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [ ] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [ ] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

### CODE REVIEW

**Review by GalaxySciTech — CHANGES_REQUESTED** [review]
leave here , i will change the code

**Review by GalaxySciTech — APPROVED** [review]

**Comment by wjrjerome** [comment]
Same as this PR. Please DO NOT merge the PR for the time being while we figure out how we going to approach the release of sol changes

**Comment by gzliudan** [comment]
I oppose modifying the contract code unless we can redeploy it.

File: contracts/validator/contract/XDCValidator.sol [hunk 1]
```diff
@@ -218,6 +218,15 @@ contract XDCValidator {
         candidateCount = candidateCount.sub(1);
         for (uint256 i = 0; i < candidates.length; i++) {
             if (candidates[i] == _candidate) {
+                for (uint256 j = 0; j < ownerToCandidate[msg.sender].length; j++) {
+                  if (ownerToCandidate[msg.sender][j] == candidates[i]){
+                    delete ownerToCandidate[msg.sender][j];
+                  }
+                }
+                ownerToCandidate[msg.sender] = removeZeroAddresses(ownerToCandidate[msg.sender]);
```
Comment (gzliudan) [inline contracts/validator/contract/XDCValidator.sol:226]: The below codes will save gas:

```solidity
                address[] memory cacheArray = ownerToCandidate[msg.sender];
                uint256 cacheArrayLength = cacheArray.length;
                for (uint256 j = 0; j < cacheArrayLength; j++) {
                    if (cacheArray[j] == _candidate){
                        delete cacheArray[j];
                    }
                }
                ownerToCandidate[msg.sender] = removeZeroAddresses(cacheArray);
```

Comment (GalaxySciTech) [reply contracts/validator/contract/XDCValidator.sol:226]: cacheArray cant change the data from ownerToCandidate[msg.sender] , it is memory ?

Comment (GalaxySciTech) [reply contracts/validator/contract/XDCValidator.sol:226]: delete cacheArray[j] will not change ownerToCandidate[msg.sender]

**Pull request closed without merging**